nem.c

```diff
@@ -0,0 +1,590 @@
+/**
+ * Copyright (c) 2017 Saleem Rashid
+ *
+ * Permission is hereby granted, free of charge, to any person obtaining
+ * a copy of this software and associated documentation files (the "Software"),
+ * to deal in the Software without restriction, including without limitation
+ * the rights to use, copy, modify, merge, publish, distribute, sublicense,
+ * and/or sell copies of the Software, and to permit persons to whom the
+ * Software is furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included
+ * in all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, E1PRESS
+ * OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL
+ * THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES
+ * OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE,
+ * ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR
+ * OTHER DEALINGS IN THE SOFTWARE.
+ */
+
+#include "nem.h"
+
+#include <string.h>
+
+#include "base32.h"
+#include "ed25519-keccak.h"
+#include "macros.h"
+#include "ripemd160.h"
+#include "sha3.h"
+
+static inline void nem_write_u32(nem_transaction_ctx *ctx, uint32_t data) {
+	ctx->buffer[ctx->offset++] = (data >>  0) & 0xff;
+	ctx->buffer[ctx->offset++] = (data >>  8) & 0xff;
+	ctx->buffer[ctx->offset++] = (data >> 16) & 0xff;
+	ctx->buffer[ctx->offset++] = (data >> 24) & 0xff;
+}
+
+static inline void nem_write_u64(nem_transaction_ctx *ctx, uint64_t data) {
+	nem_write_u32(ctx, (data >>  0) & 0xffffffff);
+	nem_write_u32(ctx, (data >> 32) & 0xffffffff);
+}
+
+static inline void nem_write(nem_transaction_ctx *ctx, const uint8_t *data, uint32_t length) {
+	nem_write_u32(ctx, length);
+
+	memcpy(&ctx->buffer[ctx->offset], data, length);
+	ctx->offset += length;
+}
+
+static inline bool nem_can_write(nem_transaction_ctx *ctx, size_t needed) {
+	return (ctx->offset + needed) <= ctx->size;
+}
+
+static inline bool nem_write_mosaic_str(nem_transaction_ctx *ctx, const char *name, const char *value) {
+	uint32_t name_length = strlen(name);
+	uint32_t value_length = strlen(value);
+
+#define NEM_SERIALIZE \
+	serialize_u32(sizeof(uint32_t) + name_length + sizeof(uint32_t) + value_length) \
+	serialize_write((uint8_t *) name, name_length) \
+	serialize_write((uint8_t *) value, value_length)
+
+#include "nem_serialize.h"
+
+	return true;
+}
+
+static inline bool nem_write_mosaic_bool(nem_transaction_ctx *ctx, const char *name, bool value) {
+	return nem_write_mosaic_str(ctx, name, value ? "true" : "false");
+}
+
+static inline bool nem_write_mosaic_u64(nem_transaction_ctx *ctx, const char *name, uint64_t value) {
+	char buffer[21];
+
+	char *s = &buffer[sizeof(buffer) - 1];
```

Comment:
I wrote this before the `bn_format` redactor, but now we can use `bn_format_uint64` here

Reply:
Please push another commit which refactors this to use `bn_format_uint64`, thanks!

Reply:
@prusnak Fixed in 065731a